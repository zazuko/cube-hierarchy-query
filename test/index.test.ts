import { meta } from '@zazuko/vocabulary-extras-builders'
import chai, { expect } from 'chai'
import $rdf from '@zazuko/env'
import { jestSnapshotPlugin } from 'mocha-chai-jest-snapshot'
import { getHierarchy, HierarchyNode } from '../index.js'
import { ex, parse, startFuseki } from './support.js'
import { insertGeoData } from './testData.js'
import { streamClient } from './client.js'

describe('@zazuko/cube-hierarchy-query', () => {
  chai.use(jestSnapshotPlugin())

  before(startFuseki)

  before(insertGeoData)

  describe('getHierarchy', () => {
    const countriesHierarchy = parse`
      <>
        ${meta.hierarchyRoot} <Europe>, <North-America>, <South-America>, <Asia> ;
        ${meta.nextInHierarchy} <countryLevel> ;
      .
      
      <countryLevel>
        ${$rdf.ns.sh.path} [
          ${$rdf.ns.sh.inversePath} ${$rdf.ns.schema.containedInPlace} ;
        ] ;
        ${meta.nextInHierarchy} <cantonLevel> ;
      .
      
      <cantonLevel> 
        ${$rdf.ns.sh.path} [
          ${$rdf.ns.sh.inversePath} ${$rdf.ns.schema.containedInPlace} ;
        ] ;
        ${meta.nextInHierarchy} <municipalityLevel> ;
      .
      
      <municipalityLevel> 
        ${$rdf.ns.sh.path} [
          ${$rdf.ns.sh.inversePath} ${$rdf.ns.schema.containedInPlace} ;
        ] ;
        ${meta.nextInHierarchy} <districtLevel> ;
      .
      
      <districtLevel> ${$rdf.ns.sh.path} ${$rdf.ns.schema.containsPlace} .
    `

    it('loads all levels', async function () {
      // given
      const hierarchy = await countriesHierarchy

      // when
      const hierarchyQuery = getHierarchy(hierarchy.namedNode(ex('')))
      const hierarchyTree = await hierarchyQuery.execute(streamClient, $rdf)

      expect(hierarchyQuery.query.build()).to.matchSnapshot()

      const plainTree = hierarchyTree.map(toPlain)

      // then
      expect(plainTree).to.containSubset([
        {
          resource: ex('Europe'),
          name: ['Europe', 'Europa'],
          nextInHierarchy: [
            {
              resource: ex('CH'),
              name: ['Switzerland'],
              nextInHierarchy: [
                {
                  resource: ex('ZH'),
                  name: ['Kanton ZÃ¼rich'],
                  nextInHierarchy: [
                    {
                      resource: ex('Affoltern'),
                      nextInHierarchy: [
                        { resource: ex('Bonstetten'), name: 'Bonstetten' },
                        { resource: ex('Rifferswil'), name: 'Rifferswil' },
                        { resource: ex('Stallikon'), name: 'Stallikon' },
                        { resource: ex('Knonau'), name: 'Knonau' },
                      ],
                    },
                  ],
                },
                {
                  resource: ex('BE'),
                  nextInHierarchy: [
                    {
                      resource: ex('Biel-Bienne'),
                      nextInHierarchy: [
                        { resource: ex('Aegerten') },
                        { resource: ex('Ligerz') },
                        { resource: ex('Nidau') },
                        { resource: ex('Port') },
                      ],
                    },
                  ],
                },
              ],
            },
          ],
        },
        {
          resource: ex('Asia'),
        },
      ])
    })

    it('loads shallower paths', async () => {
      // given
      const hierarchy = await countriesHierarchy

      // when
      const hierarchyTree = await getHierarchy(hierarchy.namedNode(ex(''))).execute(streamClient, $rdf)
      const plainTree = hierarchyTree.map(toPlain)

      // then
      expect(plainTree).to.containSubset([{
        resource: ex('North-America'),
        nextInHierarchy: [
          { resource: ex('US') },
        ],
      },
      {
        resource: ex('South-America'),
        nextInHierarchy: [
          { resource: ex('BR') },
          { resource: ex('AR') },
          { resource: ex('VE') },
        ],
      }])
    })

    it('applies type filter', async () => {
      // given
      const hierarchy = await countriesHierarchy
      hierarchy.namedNode(ex('countryLevel'))
        .addOut($rdf.ns.sh.targetClass, ex.Country)

      // when
      const hierarchyTree = await getHierarchy(hierarchy.namedNode(ex(''))).execute(streamClient, $rdf)
      const plainTree = hierarchyTree.map(toPlain)

      // then
      expect(plainTree).not.to.containSubset([{
        resource: ex('Europe'),
        nextInHierarchy: [
          { resource: $rdf.namedNode('https://sws.geonames.org/798544') },
        ],
      }])
    })

    it('loads only chosen properties, including language filter', async () => {
      // given
      const hierarchy = await countriesHierarchy

      // when
      const hierarchyTree = await getHierarchy(hierarchy.namedNode(ex('')), {
        properties: [
          $rdf.ns.schema.identifier,
          [$rdf.ns.schema.name, { language: 'de' }],
        ],
      }).execute(streamClient, $rdf)

      // then
      const plainTree = hierarchyTree.map(toPlain)
      expect(plainTree).to.containSubset([{
        resource: ex('North-America'),
        types: [$rdf.ns.schema.Continent],
        name: ['Nordamerika'],
      }])
    })
  })
})

function toPlain(node: HierarchyNode): Record<string, unknown> {
  if (!node.nextInHierarchy.length) {
    return {
      resource: node.resource.term,
      name: node.resource.out($rdf.ns.schema.name).value,
    }
  }

  return {
    resource: node.resource.term,
    nextInHierarchy: node.nextInHierarchy.map(toPlain),
    types: node.resource.out($rdf.ns.rdf.type).terms,
    name: node.resource.out($rdf.ns.schema.name).values,
  }
}
