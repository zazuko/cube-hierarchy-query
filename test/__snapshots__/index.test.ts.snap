// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@zazuko/cube-hierarchy-query getHierarchy loads all levels 1`] = `
"PREFIX schema: <http://schema.org/>
CONSTRUCT {
  ?resource2 schema:containedInPlace ?resource1.
  ?resource1 ?resource24 ?resource25.
  ?resource3 schema:containedInPlace ?resource2.
  ?resource2 ?resource18 ?resource19.
  ?resource4 schema:containedInPlace ?resource3.
  ?resource3 ?resource13 ?resource14.
  ?resource4 schema:containsPlace ?resource5.
  ?resource4 ?resource9 ?resource10.
  ?resource5 ?resource6 ?resource7.
}
WHERE {
  {
    SELECT ?resource2 ?resource1 ?resource24 ?resource25 WHERE {
      VALUES ?resource1 {
        <http://example.com/Europe>
        <http://example.com/North-America>
        <http://example.com/South-America>
        <http://example.com/Asia>
      }
      { ?resource2 schema:containedInPlace ?resource1. }
      UNION
      { ?resource1 ?resource24 ?resource25. }
    }
  }
  UNION
  {
    SELECT ?resource3 ?resource2 ?resource18 ?resource19 WHERE {
      VALUES ?resource1 {
        <http://example.com/Europe>
        <http://example.com/North-America>
        <http://example.com/South-America>
        <http://example.com/Asia>
      }
      ?resource2 schema:containedInPlace ?resource1.
      { ?resource3 schema:containedInPlace ?resource2. }
      UNION
      { ?resource2 ?resource18 ?resource19. }
    }
  }
  UNION
  {
    SELECT ?resource4 ?resource3 ?resource13 ?resource14 WHERE {
      VALUES ?resource1 {
        <http://example.com/Europe>
        <http://example.com/North-America>
        <http://example.com/South-America>
        <http://example.com/Asia>
      }
      ?resource2 schema:containedInPlace ?resource1.
      ?resource3 schema:containedInPlace ?resource2.
      { ?resource4 schema:containedInPlace ?resource3. }
      UNION
      { ?resource3 ?resource13 ?resource14. }
    }
  }
  UNION
  {
    SELECT ?resource4 ?resource5 ?resource9 ?resource10 WHERE {
      VALUES ?resource1 {
        <http://example.com/Europe>
        <http://example.com/North-America>
        <http://example.com/South-America>
        <http://example.com/Asia>
      }
      ?resource2 schema:containedInPlace ?resource1.
      ?resource3 schema:containedInPlace ?resource2.
      ?resource4 schema:containedInPlace ?resource3.
      { ?resource4 schema:containsPlace ?resource5. }
      UNION
      {
        ?resource4 ?resource9 ?resource10.
        FILTER(!(?resource9 = schema:containsPlace))
      }
    }
  }
  UNION
  {
    SELECT ?resource5 ?resource6 ?resource7 WHERE {
      VALUES ?resource1 {
        <http://example.com/Europe>
        <http://example.com/North-America>
        <http://example.com/South-America>
        <http://example.com/Asia>
      }
      ?resource2 schema:containedInPlace ?resource1.
      ?resource3 schema:containedInPlace ?resource2.
      ?resource4 schema:containedInPlace ?resource3;
        schema:containsPlace ?resource5.
      ?resource5 ?resource6 ?resource7.
    }
  }
}"
`;
